### PR TITLE
lockstep refactoring `KernelEnergyParticles`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- rewrite kernel with the lockstep programming model
- add documentation
- indent everything within namespaces
- remove thread collective iteration over the frame list

[HowTo lockstep programming](http://picongpu.readthedocs.io/en/dev/prgpatterns/lockstep.html)

# Review

To simplify the review please add `?w=1` to the URL (will ignore all white space changes) or follow [this link](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2164/files?w=1)

# Tests

- [x] 3D KHI electrons/positrons

### CODE REVIEW

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 1]
```diff
@@ -45,294 +46,415 @@
...
+                    /* error handling: */
+                    if (!outFile)
+                    {
+                        std::cerr << "Can't open file [" << filename
+                                  << "] for output, diasble plugin output. " << std::endl;
+                        writeToFile = false;
+                    }
+
+                    /* create header of the file */
+                    outFile << "#step Ekin_Joule E_Joule" << " \n";
```
Comment (theZiz) [inline include/picongpu/plugins/EnergyParticles.hpp]: Should this code line executed if `outFile` is false? Maybe put an `else` before this line?

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 2]
```diff
@@ -45,294 +46,415 @@
...
+        std::string pluginGetName() const
+        {
+            return pluginName;
+        }
 
-    /** method to initialize plugin output and variables **/
-    void pluginLoad()
-    {
-        if (notifyPeriod > 0) /* only if plugin is called at least once */
+      /** set cell description in this plugin **/
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: format

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 3]
```diff
@@ -45,294 +46,415 @@
...
-             mapper);
-
-        dc.releaseData( ParticlesType::FrameType::getName() );
+        void restart(uint32_t restartStep, const std::string restartDirectory)
+        {
+            if( !writeToFile )
+                return;
 
-        gEnergy->deviceToHost(); /* get energy from GPU */
+            writeToFile = restoreTxtFile( outFile,
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: if you touch each line, at least make them pretty ;)

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 4]
```diff
@@ -45,294 +46,415 @@
...
+                reducedEnergy,
+                gEnergy->getHostBuffer().getBasePointer(),
+                2,
+                mpi::reduceMethods::Reduce()
+            );
+
+            /* print timestep, kinetic energy and total energy to file: */
+            if (writeToFile)
+            {
+                typedef std::numeric_limits< float_64 > dbl;
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: if you change the format of this line, pls make it `using`

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 5]
```diff
@@ -45,294 +46,415 @@
...
 
-        gEnergy->deviceToHost(); /* get energy from GPU */
+            writeToFile = restoreTxtFile( outFile,
+                                          filename,
+                                          restartStep,
+                                          restartDirectory );
+        }
 
-        float_64 reducedEnergy[2]; /* create storage for kinetic and total energy */
+        void checkpoint(uint32_t currentStep, const std::string checkpointDirectory)
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: const to the right

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 6]
```diff
@@ -45,294 +46,415 @@
...
-        AreaMapping<AREA, MappingDesc> mapper(*cellDescription);
-        /* kernel call = sum all particle energies on GPU */
-        PMACC_KERNEL(KernelEnergyParticles{})
-            (mapper.getGridDim(), block)
-            (particles->getDeviceParticlesBox(),
-             gEnergy->getDeviceBuffer().getDataBox(),
-             mapper);
-
-        dc.releaseData( ParticlesType::FrameType::getName() );
+        void restart(uint32_t restartStep, const std::string restartDirectory)
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: const to the right

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 7]
```diff
@@ -45,294 +46,415 @@
...
-        if (notifyPeriod > 0) /* only if plugin is called at least once */
+        /** method to quit plugin **/
+        void pluginUnload()
         {
-            if (writeToFile)
+            if (notifyPeriod > 0) /* only if plugin is called at least once */
             {
-                outFile.flush();
-                outFile << std::endl; /* now all data is written to file */
+                if (writeToFile)
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: bracket style

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 8]
```diff
@@ -45,294 +46,415 @@
...
+                        uint32_t const idx
+                    )
+                    {
+                        if( isParticleCtx[ idx ] )
+                        {
+                            /* get one particle */
+                            auto particle = frame[ linearIdx ];
+                            float3_X const mom = particle[ momentum_ ];
+                            // compute square of absolute momentum of the particle
+                            float_X const mom2 =
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: this is `math::abs2` :)

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 9]
```diff
@@ -23,20 +23,21 @@
 #include <mpi.h>
```
Comment (ax3l) [inline include/picongpu/plugins/EnergyParticles.hpp]: huh, we include `mpi.h` here - rather random ;)

File: include/picongpu/plugins/EnergyParticles.hpp (hunk 1 above)
Comment (psychocoderHPC) [inline include/picongpu/plugins/EnergyParticles.hpp]: you are right this line should not executed, it is a bug and I will fix it in a separate pull request

**Pull request merged**